packages/flutter_tools/lib/src/migrate/migrate_utils.dart

```diff
@@ -0,0 +1,362 @@
+      if (path.contains(':')) {
+        path = path.substring(path.indexOf(':') + 1);
+      }
+      cmdArgs.add('":(exclude)$path"');
+    } else {
+      cmdArgs.add("':(exclude)${_fileSystem.path.join(workingDirectory, kDefaultMigrateWorkingDirectoryName)}'");
+    }
+    final RunResult result = await _processUtils.run(cmdArgs, workingDirectory: workingDirectory);
+    checkForErrors(result, allowedExitCodes: <int>[-1], commandDescription: cmdArgs.join(' '));
+    print(result.stdout);
+    print(result.stderr);
+    if (result.exitCode == 0) {
+      return false;
+    }
+    return true;
+  }
+
+  /// Returns true if the workingDirectory is a git repo.
+  Future<bool> isGitRepo(String workingDirectory) async {
+    final List<String> cmdArgs = <String>['git', 'rev-parse', '--is-inside-work-tree'];
+    final RunResult result = await _processUtils.run(cmdArgs, workingDirectory: workingDirectory);
+    checkForErrors(result, allowedExitCodes: <int>[-1], commandDescription: cmdArgs.join(' '));
+    if (result.exitCode == 0) {
+      return true;
+    }
+    return false;
+  }
+
+  /// Returns true if the file at `filePath` is covered by the `.gitignore`
+  Future<bool> isGitIgnored(String filePath, String workingDirectory) async {
+    final List<String> cmdArgs = <String>['git', 'check-ignore', filePath];
+    final RunResult result = await _processUtils.run(cmdArgs, workingDirectory: workingDirectory);
+    checkForErrors(result, allowedExitCodes: <int>[0, 1, 128], commandDescription: cmdArgs.join(' '));
+    return result.exitCode == 0;
+  }
+
+  /// Runs `flutter pub upgrade --major-revisions`.
+  Future<void> flutterPubUpgrade(String workingDirectory) async {
+    final List<String> cmdArgs = <String>['flutter', 'pub', 'upgrade', '--major-versions'];
+    final RunResult result = await _processUtils.run(cmdArgs, workingDirectory: workingDirectory, allowReentrantFlutter: true);
+    checkForErrors(result, commandDescription: cmdArgs.join(' '));
+  }
+
+  /// Runs `./gradlew tasks` in the android directory of a flutter project.
+  Future<void> gradlewTasks(String workingDirectory) async {
+    final String baseCommand = _platform.isWindows ? 'gradlew.bat' : './gradlew';
+    final List<String> cmdArgs = <String>[baseCommand, 'tasks'];
+    final RunResult result = await _processUtils.run(cmdArgs, workingDirectory: workingDirectory);
+    checkForErrors(result, commandDescription: cmdArgs.join(' '));
+  }
+
+  /// Verifies that the RunResult does not contain an error.
+  ///
+  /// If an error is detected, the error can be optionally logged or exit the tool.
+  ///
+  /// Passing -1 in allowedExitCodes means all exit codes are valid.
+  bool checkForErrors(
+    RunResult result, {
+    List<int> allowedExitCodes = const <int>[0],
+    String? commandDescription,
+    bool exit = true,
+    bool silent = false
+  }) {
+    if ((result.exitCode != 0 && !allowedExitCodes.contains(result.exitCode)) && !allowedExitCodes.contains(-1)) {
```

Comment:
So exit code 0 is always allowed, even if it is not included in `allowedExitCodes`? This seems surprising.

Reply:
Probably not, I'll remove this.